Add MockOracle for APIConsumer & tests

giovannidisiena opened the pull request:
This PR adds `MockOracle.sol` for use with `APIConsumer.sol` tests in `APIConsumer.t.sol`.

giovannidisiena committed:
forge install: forge-std

giovannidisiena committed:
add api consumer & tests

PatrickAlphaC commented on src/APIConsumer.sol:
```diff
@@ -0,0 +1,93 @@
+    event DataFullfilled(uint256 volume);
+
+    /**
+     * Network: Kovan
```
We'd probably want to remove this comment right?

giovannidisiena replied:
Will do 👍

PatrickAlphaC commented on .gas-snapshot:
```diff
@@ -0,0 +1,10 @@
+APIConsumerIntegrationTest:test_apiRequestCallback() (gas: 98186)
```
What's this file for?

giovannidisiena replied:
Ah that's an artefact from my local config running `forge snapshot` - useful to see diff in gas usage between changes when developing contracts but maybe not so relevant in this repo so will remove.

PatrickAlphaC commented on .gitmodules:
```diff
@@ -8,3 +8,6 @@
 [submodule "lib/chainlink-brownie-contracts"]
 	path = lib/chainlink-brownie-contracts
 	url = https://github.com/smartcontractkit/chainlink-brownie-contracts
+[submodule "lib/forge-std"]
```
This is awesome

PatrickAlphaC commented on src/test/APIConsumer.t.sol:
```diff
@@ -0,0 +1,77 @@
+     *************************************************************************/
+
+    function test_apiRequest() public {
+        vm.expectEmit(true, false, false, true);
+        emit ChainlinkRequested(requestId);
```
Could you explain what this is doing?

giovannidisiena replied:
Sure, so the `expectEmit` cheatcode is used to prepare an expected log with `(bool checkTopic1, bool checkTopic2, bool checkTopic3, bool checkData)` specifying which topic(s) to check, and the calldata. We then emit the event we expect to see followed by a call to the originating function.

After the call there are internal checks for if logs were emitted in the expected order with the expected topics and data (as specified by the booleans). See Foundry book [here](https://onbjerg.github.io/foundry-book/reference/cheatcodes.html?highlight=expectem#expectemit) for a complete explanation.

So in this case we are checking whether the `ChainlinkRequested` event was emitted with the correct `requestId` on call to `requestVolumeData`. If we wanted to test event emission only then it would be something like this:

```
vm.expectEmit(false, false, false, false);
emit ChainlinkRequested(requestId);
apiConsumer.requestVolumeData();
```

PatrickAlphaC commented on src/test/mocks/MockOracle.sol:
```diff
@@ -0,0 +1,194 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {LinkTokenReceiver} from "../utils/LinkTokenReceiver.sol";
```
Why do we need to have this in our repo? Do we need it in our repo?

giovannidisiena replied:
If I'm not mistaken, there isn't a `LinkTokenReceiver.sol` contract in `chainlink-brownie-contracts/contracts/src/v0.8` and the `v0.7` version is pinned using `^0.7.0` so unable to use that instead.

PatrickAlphaC commented:
Left a few comments, but all in all, this is awesome!

PatrickAlphaC commented:
Also, do you think we should use the `test.t.sol` filename convention for tests?

giovannidisiena commented:
> Left a few comments, but all in all, this is awesome!

Thanks for the quick review, and awesome -- glad you think so! Will start addressing comments and request re-review when ready.

> Also, do you think we should use the `test.t.sol` filename convention for tests?

Yes, I think this convention is pretty standard from other repos I've seen using Foundry -- so I guess we should change the other file names to reflect this?

giovannidisiena committed:
address review comments

- remove kovan reference
- remove .gas-snapshot
- move to test.t.sol convention

PatrickAlphaC commented:
Thanks again for making this!
